alica_engine/include/engine/blackboard/Blackboard.h

```diff
@@ -23,25 +68,100 @@ class BlackboardImpl
     {
-        return std::any_cast<T&>(vals.at(key));
+        try {
+            return std::any_cast<T&>(vals.at(key));
+        } catch (const std::bad_any_cast& e) {
+            throw BadBlackboardAccessType(key, getNameFromType<T>());
+        }
     }
     std::any& get(const std::string& key) { return vals.at(key); }
     const std::any& get(const std::string& key) const { return vals.at(key); }
 
-    void set(const std::string& key, const std::any& value) { vals.at(key) = value; }
+    template <class T>
+    void set(const std::string& key, const T& value)
+    {
+        YAML::Node entry = getBlackboardValueNode(key);
+        if (entry[Strings::key].as<std::string>() != key) {
+            // value is not predefined in pml file
+            vals.at(key) = value;
+        } else {
+            // value is predefined in pml file
+            std::string typeString = getNameFromType<T>();
+            if (entry[Strings::stateType].as<std::string>() == typeString) {
+                vals.at(key) = value;
+            } else {
+                // types dont match, throw exception
+                throw BlackboardTypesDontMatchException(typeString, entry[Strings::stateType].as<std::string>());
+            }
+        }
+    }
 
     bool hasValue(const std::string& key) const { return vals.count(key); }
     void removeValue(const std::string& key) { vals.erase(key); }
 
+    void initDefaultValues()
+    {
+        for (YAML::Node entry : node) {
+            YAML::Node defaultValue = entry[Strings::defaultValue];
+            if (defaultValue.Type() != YAML::NodeType::Null) {
+                std::string key = entry[Strings::key].as<std::string>();
+                std::string typeString = entry[Strings::stateType].as<std::string>();
+                if (typeString == "int64_t") {
+                    set<int64_t>(key, defaultValue.as<int64_t>());
+                } else if (typeString == "bool") {
+                    set<bool>(key, defaultValue.as<bool>());
+                } else if (typeString == "std::string") {
+                    set<std::string>(key, defaultValue.as<std::string>());
+                } else if (typeString == "double") {
+                    set<double>(key, defaultValue.as<double>());
+                } else {
+                    set<std::any>(key, defaultValue.as<std::string>());
+                }
+            }
+        }
+    }
+
     void clear() { vals.clear(); }
     bool empty() const { return vals.empty(); }
     size_t size() const { return vals.size(); }
     std::unordered_map<std::string, std::any> vals;
+    YAML::Node node;
+
+private:
+    YAML::Node getBlackboardValueNode(std::string key) const
+    {
+        for (const auto& entry : node) {
+            if (entry[Strings::key].as<std::string>() == key) {
+                return entry;
+            }
+        }
+        return YAML::Node();
+    }
+
+    template <class T>
+    std::string getNameFromType() const
```

Comment:
seems like there is a simple way to do the conversion from type to name without having an if else ladder. Bonus: supports implicit conversions like int32_t to int64_t etc.
But converting from std::string to type is not so easy.
Also can add more types easily although I can't think of too many useful types that can be added. Adding ROS types will require a little bit of tweaking, but shouldn't be too difficult

```
struct Converter {
    static constexpr const char* typeNames[] = {"bool", "int64", "double", "std::string", "std::any"};
    using Types = std::variant<bool, int64_t, double, std::string, std::any>;

    // convert from type to name as string
    template <class T>
    static constexpr const char* typeName() {
        return typeNames[Types{T{}}.index()];
    }

    // set the value given the type name as a string
    static std::any setDefaultValue(const std::string& tpname, const std::string& value) {
        static constexpr std::size_t numTypes = sizeof(typeNames) / sizeof(const char*);
        for (std::size_t i = 0; i < numTypes; ++i) {
            if (tpname == typeNames[i]) {
                auto value = makeAnyFromIndex(Types{}, std::make_index_sequence<numTypes>{}, i, value);
                // store value on the blackboard
            }
        }
        return {};
    }

    template <class T>
    struct YamlAs {
        static T as(const std::string& value) {
            // call Yaml's as<T>(value) method
        }
    };

    template <class T, std::size_t INDEX, template <class> class Parser = YamlAs>
    struct makeAnyIfEqual {
        static std::any make(std::size_t index, const std::string& value) {
            return index == INDEX ? std::any{Parser<T>::as(value)} : std::any{};
        }
    };

    template <class... Ts, std::size_t... Is>
    static std::any makeAnyFromIndex(std::variant<Ts...>, std::index_sequence<Is...>, std::size_t index, const std::string& value) {
        std::any typeValues[] = { makeAnyIfEqual<Ts, Is>::make(index, value)... };
        return typeValues[index];
    }
};
```